contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -0,0 +1,903 @@
+}
+
+void Filter::sendLocalReply(Http::Code response_code, absl::string_view body_text,
+                            std::function<void(Http::ResponseHeaderMap& headers)> modify_headers,
+                            Grpc::Status::GrpcStatus grpc_status, absl::string_view details) {
+  ENVOY_LOG(debug, "sendLocalReply, response code: {}", int(response_code));
+
+  auto& state = getProcessorState();
+  auto weak_ptr = weak_from_this();
+  state.getDispatcher().post(
+      [this, &state, weak_ptr, response_code, body_text, modify_headers, grpc_status, details] {
+        ASSERT(state.isThreadSafe());
+        // TODO: do not need lock here, since it's the work thread now.
+        Thread::ReleasableLockGuard lock(mutex_);
+        if (!weak_ptr.expired() && !has_destroyed_) {
+          lock.release();
+          sendLocalReplyInternal(response_code, body_text, modify_headers, grpc_status, details);
+        } else {
+          ENVOY_LOG(info, "golang filter has gone or destroyed in sendLocalReply");
+        }
+      });
+};
+
+void Filter::continueStatus(GolangStatus status) {
+  // TODO: skip post event to dispatcher, and return continue in the caller,
+  // when it's invoked in the current envoy thread, for better performance & latency.
+  auto& state = getProcessorState();
+  ENVOY_LOG(debug, "golang filter continue from Go, status: {}, state: {}, phase: {}", int(status),
+            state.stateStr(), state.phaseStr());
+
+  auto weak_ptr = weak_from_this();
+  state.getDispatcher().post([this, &state, weak_ptr, status] {
+    ASSERT(state.isThreadSafe());
+    // TODO: do not need lock here, since it's the work thread now.
+    Thread::ReleasableLockGuard lock(mutex_);
+    if (!weak_ptr.expired() && !has_destroyed_) {
+      lock.release();
+      continueStatusInternal(status);
+    } else {
+      ENVOY_LOG(info, "golang filter has gone or destroyed in continueStatus event");
+    }
+  });
+}
+
+absl::optional<absl::string_view> Filter::getHeader(absl::string_view key) {
+  Thread::LockGuard lock(mutex_);
+  if (has_destroyed_) {
+    ENVOY_LOG(warn, "golang filter has been destroyed");
+    return "";
+  }
+  auto& state = getProcessorState();
+  auto m = state.isProcessingHeader() ? headers_ : trailers_;
+  auto result = m->get(Http::LowerCaseString(key));
+
+  if (result.empty()) {
+    return absl::nullopt;
+  }
+  return result[0]->value().getStringView();
+}
+
+void copyHeaderMapToGo(Http::HeaderMap& m, GoString* go_strs, char* go_buf) {
+  auto i = 0;
+  m.iterate([&i, &go_strs, &go_buf](const Http::HeaderEntry& header) -> Http::HeaderMap::Iterate {
+    auto key = std::string(header.key().getStringView());
+    auto value = std::string(header.value().getStringView());
+
+    // std::cout << "idx: " << i << ", key: " << key << ", value: " << value << std::endl;
+
+    auto len = key.length();
+    go_strs[i].n = len;
+    go_strs[i].p = go_buf;
+    memcpy(go_buf, key.data(), len); // NOLINT(safe-memcpy)
```

Comment:
Why are you not use safe-memcpy here and elsewhere?

Reply:
Done. add some comments for it. Because `go_buf` is the heap memory of go, and the length is the total length of all keys and values in the header. So here use `memcpy` is safe.